### PR TITLE
Feature/pneumatics base

### PR DESCRIPTION
**Pull request by adamv17**
Add the pneumatics base and utilize it to check whether the solenoids are connected.

### CODE REVIEW

**Review by barel13 — APPROVED** [review]
Now we only need to fix this issue #37

File: src/main/java/frc/robot/Robot.java [hunk 1]
```diff
@@ -5,10 +5,7 @@
 package frc.robot;
 
 import com.kauailabs.navx.frc.AHRS;
-import edu.wpi.first.wpilibj.DriverStation;
-import edu.wpi.first.wpilibj.PowerDistribution;
-import edu.wpi.first.wpilibj.SPI;
-import edu.wpi.first.wpilibj.TimedRobot;
+import edu.wpi.first.wpilibj.*;
```
Comment (katzuv) [inline src/main/java/frc/robot/Robot.java]: Don't use star imports, they clutter your namespace. Import only what you need.

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -23,6 +20,7 @@
     public PowerDistribution pdp = new PowerDistribution();
     private RobotContainer m_robotContainer;
     private Command m_autonomousCommand;
+    public static final PneumaticsBase pneumaticsBase = PneumaticsBase.getForType(0, PneumaticsModuleType.CTREPCM);
```
Comment (katzuv) [inline src/main/java/frc/robot/Robot.java]: You should use the [`PneumaticsControlModule`](https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj/PneumaticsControlModule.html) class directly.

File: src/main/java/frc/robot/subsystems/hood/Hood.java [hunk 3]
```diff
@@ -3,14 +3,21 @@
...
 public class Hood extends SubsystemBase {
     private static Hood INSTANCE;
-    private final Solenoid angleChanger = new Solenoid(PneumaticsModuleType.CTREPCM, SOLENOID);
+    private final Solenoid angleChanger;
 
     private Hood() {
+        if (Robot.pneumaticsBase.checkSolenoidChannel(SOLENOID)) {
+            angleChanger = new Solenoid(PneumaticsModuleType.CTREPCM, SOLENOID);
+        } else {
+            angleChanger = null;
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/hood/Hood.java:20]: Wouldn't that crash the code if i.e. `close()` is called?

**Review by katzuv — CHANGES_REQUESTED** [review]
This is a great feature but I'm afraid it doesn't do what you want. The docs for `checkSolenoidChannel()` say it returns whether "a solenoid channel is valid", not if it's connected or not. "Valid" is probably in the valid range, as you can see in the implementation ([source](https://github.com/wpilibsuite/allwpilib/blob/10f19e6fc3ba982b08bb31da8c21c7b8241d0502/hal/src/main/native/sim/CTREPCM.cpp#L76-L78)):
```cpp
HAL_Bool HAL_CheckCTREPCMSolenoidChannel(int32_t channel) {
  return channel < kNumCTRESolenoidChannels && channel >= 0;
}
```

File: src/main/java/frc/robot/subsystems/hood/Hood.java (hunk 3 above)
Comment (adamv17) [reply src/main/java/frc/robot/subsystems/hood/Hood.java:20]: We wouldn't use the close it just simplifies the testing process if something is not connected yet

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/hood/Hood.java:20]: Will the testing be added in another branch?

File: src/main/java/frc/robot/Robot.java [hunk 4]
```diff
@@ -23,6 +24,7 @@
     public PowerDistribution pdp = new PowerDistribution();
     private RobotContainer m_robotContainer;
     private Command m_autonomousCommand;
+    public static final PneumaticsControlModule pneumaticsBase = new PneumaticsControlModule(0);
```
Comment (katzuv) [inline src/main/java/frc/robot/Robot.java:27]: 1. The name now doesn't match, use either `PneumaticsControlModule` or `pcm`.
2. The default ID is `0` so you can use the simpler constructor: 
```suggestion
    public static final PneumaticsControlModule pneumaticsBase = new PneumaticsControlModule();
```

File: src/main/java/frc/robot/subsystems/conveyor/Conveyor.java [hunk 5]
```diff
@@ -39,6 +42,11 @@ private Conveyor() {
         colorMatch.addColorMatch(NONE);
         cargoPositions.add(DriverStation.Alliance.Invalid.name());
         cargoPositions.add(DriverStation.Alliance.Invalid.name());
+        if (Utils.isBitSet(Robot.pneumaticsBase.getSolenoids(), SOLENOID)) {
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:45]: `getSolenoids()` is used in `Solenoid.get()` ([source](https://github.com/wpilibsuite/allwpilib/blob/5635f33a328144542e4ca7947c8c54614ce30de3/wpilibc/src/main/native/cpp/Solenoid.cpp#L50-L53)):
```cpp
bool Solenoid::Get() const {
  int currentAll = m_module->GetSolenoids();
  return (currentAll & m_mask) != 0;
}
```
So I think that unfortunately, that doesn't help either (or if it does, you can simply call `Solenoid`'s `get()`). What does `Solenoid.get()` return for a disconnected solenoid?

Also, just a quick question: Aren't double solenoids used in the robot?

BTW, I thought of having a "general test command" which you can run in the roboRIO's Test mode. This command will be a command group with basic commands for each subsystem (even use the BIT's one). This way, you can see whether mechanisms behave in the way you expect them to.
You can also do some more complicated tests. For example, the command for the shooter will print an error if the shooter didn't reach its setpoint 5 seconds after it started moving.
Moreover, think about getting user input and having that test command handle a "TODO list" for you. That way you don't need to remember what checks need to be done and which of these have finished.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/conveyor/Conveyor.java (hunk 5 above)
Comment (adamv17) [reply src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:45]: 1. Honestly I am not sure anymore
2. We use single solenoids
3. Yes we will add bits for each subsystem, how would you handle a TODO list?

**Pull request closed without merging**